runtime/src/main/java/org/corfudb/runtime/DistributedCompactor.java

```diff
@@ -0,0 +1,558 @@
+
+    public static final String CHECKPOINT = "checkpoint";
+
+    public static final StringKey COMPACTION_MANAGER_KEY = StringKey.newBuilder().setKey("CompactionManagerKey").build();
+    public static final StringKey CHECKPOINT_KEY = StringKey.newBuilder().setKey("minCheckpointToken").build();
+    public static final StringKey UPGRADE_KEY = StringKey.newBuilder().setKey("UpgradeKey").build();
+
+    private CorfuStore corfuStore = null;
+    private Table<StringKey, CheckpointingStatus, Message> compactionManagerTable = null;
+    private Table<TableName, CheckpointingStatus, Message> checkpointingStatusTable = null;
+    private Table<TableName, ActiveCPStreamMsg, Message> activeCheckpointsTable = null;
+    private Table<StringKey, TokenMsg, Message> checkpointTable = null;
+
+    private CheckpointLivenessUpdater livenessUpdater;
+
+    public static class CorfuTableNamePair {
+        public TableName tableName;
+        public CorfuTable corfuTable;
+
+        public CorfuTableNamePair(TableName tableName, CorfuTable corfuTable) {
+            this.tableName = tableName;
+            this.corfuTable = corfuTable;
+        }
+    }
+
+    public DistributedCompactor(CorfuRuntime corfuRuntime, CorfuRuntime cpRuntime, Optional<String> persistedCacheRoot) {
+        this.runtime = corfuRuntime;
+        this.cpRuntime = cpRuntime;
+        this.persistedCacheRoot = persistedCacheRoot;
+        this.isClient = false;
+        this.clientName = corfuRuntime.getParameters().getClientName();
+    }
+
+    public DistributedCompactor(CorfuRuntime corfuRuntime) {
+        this.runtime = corfuRuntime;
+        this.cpRuntime = null;
+        this.persistedCacheRoot = Optional.empty();
+        this.isClient = true;
+        this.clientName = corfuRuntime.getParameters().getClientName();
+    }
+
+    private void openCheckpointingMetadataTables() throws Exception {
+        if (this.corfuStore != null) { // only run this method once
+            return;
+        }
+        this.corfuStore = new CorfuStore(this.runtime);
+        livenessUpdater = new CheckpointLivenessUpdater(corfuStore);
+        log.debug("Opening all the checkpoint metadata tables");
+        this.compactionManagerTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME,
+                StringKey.class,
+                CheckpointingStatus.class,
+                null,
+                TableOptions.fromProtoSchema(CheckpointingStatus.class));
+
+        this.checkpointingStatusTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                DistributedCompactor.CHECKPOINT_STATUS_TABLE_NAME,
+                TableName.class,
+                CheckpointingStatus.class,
+                null,
+                TableOptions.fromProtoSchema(CheckpointingStatus.class));
+
+        this.activeCheckpointsTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
```

Comment:
Line 121 `CheckpointLivenessUpdater` already opens this table. Consider moving line 121 after all openTables and let CheckpointLivenessUpdater call corfuStore.getTable instead of openTable.